### ISSUE TITLE
task: Train and test text2semantic under decoder only framework for ichigo v0.5

### ISSUE THREAD

**Issue by tikikun** [janhq/ichigo#145]
# Motivation
Since ichigo v0.5 will support additional language that will make the traditional t2s obsolete. This is a good chance to introduce a t2s framework that we have full control over.

# Goal
 Be able to handle any arbitrary language
- Current Ichigo approach (WhisperVQ) is only trained for 7 languages
- We cannot find a ASR or STT module for every new language
- Our approach: Text to Semantic (same semantic space as Ichigo speech embeddings)

# Methodology
1. Teacher: Speech -> semantic (Whisper encoder)
2. Student: Text -> semantic (decoder, Qwen base model)

First step: Run base case first, with English first (before exploring other languages)

- WhisperSpeech text-to-semantic model failed for our synthetic data pipeline, producing incompatible sound tokens that disrupted Ichigo’s comprehension.
  - Proposed solution: a custom decoder-only text-to-semantic model (<3B parameters, similar to Qwen 2.5) with knowledge transfer from WhisperVQ and a more efficient architecture.
  - Processed 10k English samples from MLS Eng 10k dataset (2.42M samples) using WhisperVQ for semantic token extraction, adding a <|text_to_semantic|> task token in user turns.
  - Example dataset: [Instruction data](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq).
- Modified Qwen 2.5 0.5B model:
  - Introduced `<|text_to_semantic|>` task token and added 512 sound tokens + 3 special tokens (start, end, mask) to its vocabulary.
  - Trained with instruction-based samples for text-to-semantic conversion.
  - Embedded control tokens without modifying the tokenizer, scaled embedding layer and LM head to `[152,192](https://github.com/QwenLM/Qwen/issues/419)` tokens for training speed optimization.

# What needed to be done:
-  [x] Train t2s on a decoder model to test feasibility (we already have the data) under ichigo v0.4
-  [ ] Train t2s on decoder model on new quantizer after janhq/WhisperSpeech#2 is done

## Experiments 

| Run ID | Date       | Model Config | Dataset | Learning Rate | Batch Size | Steps | Loss | Hardware | 
|--------|------------|--------------|---------|---------------|------------|-------|------|----------|
| exp-t2s-0.5B | 2024-11-28 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-0.5B.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq) | 1e-3 | 96  | 28810 | [1.6-1.7](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-0.5B.png) | ~ 4 hours on 2xH100 | 
| exp-t2s-1.5B-1 | 2024-11-29 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-1.5B-1.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq) | 1e-3 | 84  | 28810 | [2.64](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-1.5B-1.png) | ~ 10 hours on 6xA6000 | 
| exp-t2s-1.5B-2 | 2024-11-30 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-1.5B-2.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq) | 1e-4 | 84  | 28810 | [1.84](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-1.5B-2.png) | ~ 10 hours on 6xA6000 |
| exp-t2s-llama3.2-1B | 2024-12-1 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-llama3.2-1B.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq-llama3.2) | 1e-4 | 96  | 25208 | [1.73](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-llama3.2-1B.png) | ~ 9 hours on 6xA6000 |  
| exp-t2s-llama3.2-1B-1 | 2024-12-2 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-llama3.2-1B.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq-llama3.2) | 1.5e-4 | 192  | 12604 | [1.77](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-llama3.2-1B-1.png) | ~ 6 hours on 6xA6000 |  
| exp-t2s-llama3.2-1B-2 | 2024-12-3 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-1.5B-2.yaml) | [Instruction text to sound sementic token](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq-llama3.2) | 1.5e-4 | 192  | 57930 | [1.44](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-llama3.2-1B-2.png) | ~30 hours on 6xA6000 |  
| exp-t2s-llama3.2-1B-dedup | 2024-12-5 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-llama3.2-1B-dedup.yaml) | [Instruction deduplicated sound sementic token](https://huggingface.co/datasets/jan-hq/instruction-convert-audio-whispervq-llama3.2-dedup) | 1.5e-4 | 168  | 33000 | [1.73](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-llama3.2-1B-dedup.png) | ~30 hours on 6xA6000 | 
| exp-t2s-llama3.2-1B-compress | 2024-12-9 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-llama3.2-1B-dedup.yaml) | [Instruction compressed sound sementic token](https://huggingface.co/datasets/jan-hq/instruction-convert-audio-whispervq-llama3.2-compress) | 3e-4 | 168  | 44713 | [1.6-1.63](https://github.com/janhq/research/blob/main/assets/T2S-Exp/loss/exp-t2s-llama3.2-1B-compress.png) | ~33 hours on 6xA6000 | 
| exp-t2s-llama3.2-1B-compress-1 | 2024-12-10 | [Full-Finetune](https://github.com/janhq/research/blob/main/assets/T2S-Exp/config/exp-t2s-llama3.2-1B-dedup.yaml) | [Instruction compressed sound sementic token](https://huggingface.co/datasets/jan-hq/instruction-convert-audio-whispervq-llama3.2-compress) | 3e-4 | 168  | Updated soon | Updated soon | ~33 hours on 6xA6000 | 

## Inference using different repetition settings
| Experiment ID | Inference Result | Note |
|--------------|-----|------|
| exp-t2s-llama3.2-1B-1 | [exp-t2s-llama3.2-1B-1-result](https://docs.google.com/spreadsheets/d/1nzDHBOcqxDZFNs73bQkw_wZlWJ-4cf-dXg1TJXmnCCU/edit?usp=sharing) | Model don't know who to generate <sound_end> token, have repetitive and hallucination problem. |
| exp-t2s-llama3.2-1B-2 | [exp-t2s-llama3.2-1B-2](https://docs.google.com/spreadsheets/d/1KaI1dsHAUxUm4gm3KfG1cEmC_jfEEDK0zKg6mB8ZyWs/edit?gid=219322134#gid=219322134)  | Successfully train a Text to Semantic model, model can generate the sound token that Ichigo v0.4 can understand |

**Comment by tikikun** [comment]
> # Idea: Add duration tokens
> Observations:
> 
> * Loss did not converge as well after we de-dup, something must be lost
> * Repetition is the only difference, how many time a sound token is repeated
> * Continuous embedding is a by product of how Whisper Encoder process (each embedding is slightly different from other one even though they are almost the same)
> 
> Theories:
> 
> * We need to be able to "add back the "repetition" information
> * After we make sure the information is kept the same, the loss behaviour should be relatively the same.
> 
> Implementation:
> 
> * We create "<|dur_xxx|> (with x is an integer) to represent the repetition inside each repeated sound tokens group
> 
> ![Image](https://private-user-images.githubusercontent.com/22268502/393671619-a43984ee-1dbc-4f44-aad3-94b5f7cfed5c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4ODUyMTQsIm5iZiI6MTczMzg4NDkxNCwicGF0aCI6Ii8yMjI2ODUwMi8zOTM2NzE2MTktYTQzOTg0ZWUtMWRiYy00ZjQ0LWFhZDMtOTRiNWY3Y2ZlZDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjExVDAyNDE1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmNjNkNjdjMGY0MTc0M2JkNzJlOTdmNWJjNThhNTMzNWVhNmUzMTRmOWEzMTkyMWI4ZjgyYzQzNDIzYjRiMmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Iq90NREsz46QgwvnQbIWhU_b5XCNwAacpAMdZ9khjeA)
> 
> # Extra Information
> Why some word might result in repetition?
> 
> At first glance it's tempting to think that the same information in a repetitive token (or embedding) might be redundant. But if we take a closer look at long and short vowels in English it might not be the case
> 
> ### Example:
> * Sheep - i
> * Ship - ɪ
> 
> The only way for you to discern the difference between Sheep and Ship sometimes in English speaking is only whether the i sound is long or short, or the **duration** part of it.
> 
> By de-duplicating the duration, **everything becomes a short sound** but your target training still **sheep or sip** (either long or short) make it impossible to really converge.
> 
> Hence, **duration** is the information that is left out when you de-dup.

This idea needs further validation on different generation len

**Comment by tikikun** [comment]
@bachvudinh please add validation on longer sequence

**Comment by bachvudinh** [comment]
- Our data length distribution. 
   ![image](https://github.com/user-attachments/assets/e2af3c22-7921-4803-9e78-62ae4f253c22)

**Comment by bachvudinh** [comment]
![image](https://github.com/user-attachments/assets/00f599e8-e818-417b-bb81-67ae01414707)

**Comment by PodsAreAllYouNeed** [comment]
A good point of reference for the results for the text-to-semantic model would be the F5-TTS model. 

We can also use their methodology of Librispeech-PC test-clean to test. Also, we can send the text into F5-TTS to generate, then take the generated audio and put it back into the full whisperVQ+Decoder model for ASR, as an additional point of reference. (This only needs to be done once for each time we retrain the VQ)

<img width="787" alt="Image" src="https://github.com/user-attachments/assets/aafe0bcc-d2bb-4b8b-b63a-fef313337dd1" />

Looking at the table from F5-TTS, it seems that a good target for the WER of the t2s -> decoder round-trip would be 2-3% on the LibriSpeech test-clean dataset.

**Comment by bachvudinh** [comment]
![image](https://github.com/user-attachments/assets/590e31e5-4561-418e-a12d-dcc9a9a1975e)

**Comment by PodsAreAllYouNeed** [comment]
i used stats.csv to plot a histogram of the delta between t2s and whispervq

<img width="490" alt="image" src="https://github.com/user-attachments/assets/710c3c1d-9b9d-4926-a40b-b18accf3475b" />

The WER difference between t2s and whisperVQ is very small, in a majority of the samples, the delta it within 0.5WER. This is a very good sign that our t2s model is behaving in a similar manner to the whisper encoder

**Comment by FerLuisxd** [comment]
Maybe take a look at https://huggingface.co/FunAudioLLM/SenseVoiceSmall as an alternative? It claims to be multilinguan (50 languages) but also not, could be worth a try as it is small and accurate apparently.

**Comment by hahuyhoang411** [comment]
image for comparing speed of speechless vs old pipeline

<img width="426" alt="Image" src="https://github.com/user-attachments/assets/9c42282a-648e-4552-b25c-a12269e364ec" />
